### PR TITLE
[AutoDiff] [stdlib] Reimplement Array's Differentiable conformance and fix bugs.

### PR DESCRIPTION
**Pull request by rxwei**
* Simplify `Array`'s `Differentiable` conformance by reimplementing it with a nested `TangentVector` struct instead of `DifferentiableView`. This is possible because `Differentiable.AllDifferentiableVariable` has been removed.
* Fix a bug in `SILGenFunction::getThunkedAutoDiffLinearMap` where there's a previously unexercised code path that creates an ownership-unqualified `load` in an OSSA function.
* Fix a bug in the differentiation transform where linear map structs are lowered to address types under resilience. Linear map structs should always be `@frozen` and loadable.
* Rename `addFixedLayoutAttr` to `addFrozenAttrIfApplicable` because `@frozen` is the official attribute. We also want to remove this sometime since it does not exist in master and is only used in two places in differentiation.

### CODE REVIEW

File: stdlib/public/core/Array.swift [hunk 1]
```diff
@@ -1913,183 +1913,110 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
...
-          TangentVector(Array<Element.TangentVector>(
-            gradientIn.base[lhs.count...])))
-      }
-      return (lhs + rhs, pullback)
+  @inlinable
+  static func _vjpConcat(
+    _ lhs: [Element], _ rhs: [Element]
+  ) -> ([Element], (TangentVector) -> (TangentVector, TangentVector)) {
+    func pullback(_ v: TangentVector) -> (TangentVector, TangentVector) {
+      precondition(v.elements.count == lhs.count + rhs.count)
```
Comment (bartchr808) [inline stdlib/public/core/Array.swift]: Why did you not keep the precondition string `... should receive gradient with count equal to sum of...` in the precondition?

File: stdlib/public/core/Array.swift [hunk 2]
```diff
@@ -1913,183 +1913,110 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
...
     return (self[index], pullback)
   }
 
-  public static func _vjpPlus(_ lhs: [Element], _ rhs: [Element]) ->
-    ([Element], (TangentVector) -> (TangentVector, TangentVector)) {
-      func pullback(_ gradientIn: TangentVector) ->
-        (TangentVector, TangentVector) {
-        precondition(
-          gradientIn.base.count == lhs.count + rhs.count,
-          "+ should receive gradient with count equal to sum of operand " +
```
Comment (bartchr808) [inline stdlib/public/core/Array.swift]: Why not keep this extra precondition description?

Comment (rxwei) [reply stdlib/public/core/Array.swift]: It is kept on line 2014. The error message is not super user-facing IMO, so I removed it for now.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -5995,9 +6000,11 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
     auto addArithConf = swiftModule->lookupConformance(astType, addArithProto);
     assert(addArithConf.hasValue() &&
            "Missing conformance to `AdditiveArithmetic`");
-    auto subMap =
-        SubstitutionMap::get(genericSig, {astType}, {*addArithConf, *diffConf});
-    builder.createApply(ai->getLoc(), fnRef, subMap,
+//    auto subMap =
+//        SubstitutionMap::get(genericSig, {astType}, {*addArithConf, *diffConf});
```
Comment (bartchr808) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Commented code could remove

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done

**Comment by rxwei** [comment]
Note: Need to fix errors on the TensorFlow side.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by dan-zheng** [comment]
I accidentally deleted `tensorflow` branch, which closed this PR. That was not intentional, sorry!
It would be nice to protect `tensorflow` branch against deletion to prevent this from happening again.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**